rustls/src/crypto/aws_lc_rs/hpke.rs

```diff
@@ -0,0 +1,774 @@
+use alloc::boxed::Box;
+use alloc::sync::Arc;
+use alloc::vec;
+use alloc::vec::Vec;
+use core::fmt::{self, Debug, Formatter};
+
+use aws_lc_rs::aead::{
+    Aad, BoundKey, Nonce as AwsNonce, NonceSequence, OpeningKey, SealingKey, UnboundKey,
+};
+use aws_lc_rs::{aead, agreement, hkdf, hmac};
+
+use crate::crypto::aws_lc_rs::tls13::RingHkdf;
+use crate::crypto::hpke::{
+    EncapsulatedSecret, Hpke, HpkeOpener, HpkePrivateKey, HpkeProvider, HpkePublicKey, HpkeSealer,
+    HpkeSuite as HpkeSuiteId,
+};
+use crate::crypto::tls13::{Hkdf, HkdfExpander, OutputLengthError};
+use crate::msgs::enums::{HpkeAead as HpkeAeadId, HpkeKdf as HpkeKdfId, HpkeKem as HpkeKemId};
+use crate::msgs::handshake::HpkeSymmetricCipherSuite;
+use crate::{Error, OtherError};
+
+/// Default [RFC 9180] Hybrid Public Key Encryption (HPKE) provider backed by aws-lc-rs cryptography.
+///
+/// Enables [ALL_SUPPORTED_SUITES]. To customize which suites are supported, construct
+/// your own [Provider], customizing the `suites` as required.
+///
+/// [RFC 9180]: https://www.rfc-editor.org/rfc/rfc9180.html
+pub static DEFAULT_PROVIDER: &'static dyn HpkeProvider = &Provider {
+    suites: ALL_SUPPORTED_SUITES,
+};
+
+/// All supported HPKE suites.
+pub static ALL_SUPPORTED_SUITES: &[&Suite] = &[
+    DHKEM_P256_HKDF_SHA256_AES_128,
+    DHKEM_P256_HKDF_SHA256_AES_256,
+    DHKEM_P256_HKDF_SHA256_CHACHA20_POLY1305,
+    DHKEM_P384_HKDF_SHA384_AES_128,
+    DHKEM_P384_HKDF_SHA384_AES_256,
+    DHKEM_P384_HKDF_SHA384_CHACHA20_POLY1305,
+    DHKEM_P521_HKDF_SHA512_AES_128,
+    DHKEM_P521_HKDF_SHA512_AES_256,
+    DHKEM_P521_HKDF_SHA512_CHACHA20_POLY1305,
+    DHKEM_X25519_HKDF_SHA256_AES_128,
+    DHKEM_X25519_HKDF_SHA256_AES_256,
+    DHKEM_X25519_HKDF_SHA256_CHACHA20_POLY1305,
+];
+
+/// A [RFC 9180] Hybrid Public Key Encryption (HPKE) provider backed by aws-lc-rs cryptography.
+///
+/// [RFC 9180]: https://www.rfc-editor.org/rfc/rfc9180.html
+#[derive(Debug)]
+pub struct Provider {
+    /// HPKE algorithm suites supported by the provider.
+    pub suites: &'static [&'static Suite],
+}
+
+impl Provider {
+    fn suite(&self, id: HpkeSuiteId) -> Option<&'static Suite> {
+        self.suites
+            .iter()
+            .find(|s| s.suite == id)
+            .copied()
+    }
+}
+
+impl HpkeProvider for Provider {
+    fn start(&self, id: &HpkeSuiteId) -> Result<Box<dyn Hpke + 'static>, Error> {
+        Ok(Box::new(
+            self.suite(*id)
+                .ok_or(Error::General("unsupported HPKE suite".into()))?,
+        ))
+    }
+
+    fn supports_suite(&self, id: &HpkeSuiteId) -> bool {
+        self.suite(*id).is_some()
+    }
```

Comment:
I think they're more useful for the entry-point to a user of HPKE, like in the ECH PR and so seem underutilized here. 

I think we do want to have some mechanism for selecting the `dyn Hpke` that doesn't need to be repeated by each use-case. Maybe it could be lifted onto the `CryptoProvider` if we end up moving the supported suites there?